src/Chainweb/Pact/PactService.hs

```diff
@@ -681,23 +695,34 @@ withCheckpointerWithoutRewind target caller act = do
     unlessM ((<= 1) <$> asks _psCheckpointerDepth) $ do
         error $ "Code invariant violation: to many nested calls of withCheckpointerRewind. Please report this as a bug."
 
+    case target of
+        Just h -> setParentHeader (caller <> ".withCheckpointerWithoutRewind") h
```

Comment:
We seem to have outgrown `psParentHeader` at this point, especially since `withCurrentCheckpointer` hits the db first. I think we're ready to move it into the environment and out of state, as there is never a case where we invoke PactService and expect it to be set properly already. 

I propose it should be a `Maybe ParentHeader` in `PactServiceEnv`, and be `Nothing` outside of any `local` call (ie in the `serviceRequests` loop), so that code that needs to _read_ it (`execTransactions` etc) will simply fail if it's not set. Then we'll finally get out from under the state problem.

What do you think of including this change in this PR?

Reply:
Yes, I think, that's the right approach. However, currently, there's one situation when `_psParentHeader` is used from the state: when `withCurrentCheckpointer` is used and the the last saved block isn't yet available in the block header db. 

This can happen, for instance, when an `execLocal` request immediately follows an `execValidateBlock` request. In those situations the '_psParentHeader' is used and it would be inefficient to restore the checkpointer to the previous block, because any subsequent new block or validation request is most likely to happen on top of the current header, which would have to be replayed if we would restore the checkpointer.

One can argue, that it would be the job of the checkpointer to remember the last block header internally. I can see good arguments both way. There is also `_psStateValidated :: Maybe BlockHeader`, which seems unused. Maybe this was used originally for this purpose? I was about it remove it, but maybe I should leave it and instead use it to initialize '_psParentHeader' in 'withCurrentCheckpointer'?

Reply:
I propose the following:

1. Implement your proposal and move '_psParentHeader' into the environment. That would give us the benefits that you describe.

2. solve the issue of the pending recently validated block header by using the existing '_psStateValidated' component in pact service state. It seems that it gets already initialized with the correct block header. It would be used for the single purpose of providing the block header of the latest block for initializing `_psParentHeader` in `withCurrentCheckpointer`.

Given that this PR is already pretty big and also has already been tested extensively, I would propose to implement that in a follow up PR.

Reply:
I'm a little apprehensive about resurrecting `_psStateValidated` as it's currently obsolete, so we don't know how correct its maintenance is (which is the problem we're going to get rid of with `_psParentHeader`). 

Also it seems like `withCurrentCheckpointer` could simply grab the latest block in checkpointer (via `_cpGetLatestBlock`) and restore with that, it's a miniscule db (and probably in sqlite RAM anyway) hit given that `local` and pre-insert have to hit the db anyway.

I would much prefer less state + more correctness, which would mean deleting `_psStateValidated` and proceeding with #1 above.